src/test3/test-3-solution.tsx

```diff
@@ -0,0 +1,223 @@
+    { id: 5, name: "Ray-Ban Sunglasses", category: "Accessories", price: 150, age: 21 },
+    { id: 6, name: "Uniqlo T-Shirt", category: "Clothes", price: 30, age: 16 },
+];
+
+function getProducts() {
+    return Promise.resolve(products);
+}
+
+const stateContext = createContext<{
+    products: Product[];
+    filtered: Product[];
+    search: string;
+    category: string;
+    priceRange: [number, number];
+    ageRange: [number, number];
+}>({
+    products: [],
+    filtered: [],
+    search: "",
+    category: "",
+    priceRange: [0, 500],
+    ageRange: [0, 100],
+});
+
+const dispatchContext = createContext((payload: {
+    type: string;
+    payload: any;
+}) => { });
+
+
+function Provider({
+    children
+}: {
+    children: React.ReactNode;
+}) {
+    const [state, dispatch] = useReducer((state, action) => {
+        switch (action.type) {
+            case "SET_PRODUCTS":
+                return { ...state, products: action.payload };
+            case "SET_FILTERED":
+                return { ...state, filtered: action.payload };
+            case "SET_SEARCH":
+                if (action.payload === "") {
+                    return { ...state, search: action.payload, filtered: state.products };
+                }
+                const filtered = state.filtered.filter(item => item.name.toLowerCase().includes(action.payload.toLowerCase()));
+                return { ...state, search: action.payload, filtered };
+            case "SET_CATEGORY":
+                if (action.payload === "") {
+                    return { ...state, category: action.payload, filtered: state.products };
+                }
+                const filteredByCategory = state.products.filter(item => item.category === action.payload);
+                return { ...state, category: action.payload, filtered: filteredByCategory };
+            case "SET_PRICE_RANGE":
+                const [minPrice, maxPrice] = action.payload;
+                const filteredByPrice = state.products.filter(item => item.price >= minPrice && item.price <= maxPrice);
+                return { ...state, priceRange: action.payload, filtered: filteredByPrice };
+            case "SET_AGE_RANGE":
+                const [minAge, maxAge] = action.payload;
+                const filteredByAge = state.products.filter(item => item.age >= minAge && item.age <= maxAge);
+                return { ...state, ageRange: action.payload, filtered: filteredByAge };
+            default:
+                return state;
+        }
+    }, {
+        products: [] as Product[],
+        filtered: [] as Product[],
+        search: "",
+        category: "",
+        priceRange: [0, 500] as [number, number],
+        ageRange: [0, 100] as [number, number],
+    });
+
+    useEffect(() => {
+        getProducts().then((data) => {
+            dispatch({ type: "SET_PRODUCTS", payload: data });
+            dispatch({ type: "SET_FILTERED", payload: data });
```

Comment:
Please do not try to follow storing 2 different data like products and filtered products, which violates:
- single source of truth -> lack of maintainability
- sparse filter logic accross SET_FILTER logics
You might think of trade-off about performance when doing filter, but we have many ways to reduce it over the maintainability